diff --git src/sysdep.c src/sysdep.c
index d100a5cb50..92104ed57d 100644
--- src/sysdep.c
+++ src/sysdep.c
@@ -1818,8 +1818,7 @@ handle_arith_signal (int sig)
 
 /* Alternate stack used by SIGSEGV handler below.  */
 
-static unsigned char sigsegv_stack[SIGSTKSZ];
-
+static max_align_t sigsegv_stack[ ( 64 * 1024 + sizeof( max_align_t ) - 1 ) / sizeof( max_align_t ) ];
 
 /* Return true if SIGINFO indicates a stack overflow.  */
 
